### PR TITLE
Waive all fees collected on refunded donations

### PR DESCRIPTION
**Pull request by sampoder**
WIP because I want to test, what'd be a good way of testing?

### CODE REVIEW

**Comment by cjdenio** [comment]
Do we want to do this from an operational standpoint? Stripe doesn't waive fees when we refund a payment

**Comment by sampoder** [comment]
Not sure, there was an open issue, I was playing around

**Comment by rluodev** [comment]
@hackliv Would this be something that we'd want to do?

**Comment by sampoder** [comment]
For reference, here's the issue #2093

**Comment by hackliv** [comment]
If there is a way to do this automatically, then yes, would be great for Ops to not have to worry about!

**Comment by hackliv** [comment]
I will share though that I'd prefer prioritizing instant fee waive/un-waive on pending disbursements over this though if you all are starting a dive into bug fixes for fees!

File: app/services/donation_service/refund.rb [hunk 1]
```diff
@@ -15,10 +15,19 @@ def run
         # 2. Un-front all pending transaction associated with this donation
         donation.canonical_pending_transactions.update_all(fronted: false)
 
-        # 3. Process remotely
+        # 3. Waive all fees collected
+
+        donation.canonical_transactions.each do |ct|
+          fee = ct.canonical_event_mapping.fees.first
```
Comment (garyhtou) [inline app/services/donation_service/refund.rb]: this line may need to be updated due to https://github.com/hackclub/hcb/pull/4773

File: app/services/donation_service/refund.rb [hunk 2]
```diff
@@ -15,10 +15,19 @@ def run
         # 2. Un-front all pending transaction associated with this donation
         donation.canonical_pending_transactions.update_all(fronted: false)
 
-        # 3. Process remotely
+        # 3. Waive all fees collected
+
```
Comment (garyhtou) [inline app/services/donation_service/refund.rb]:
```suggestion
```

File: app/services/donation_service/refund.rb [hunk 3]
```diff
@@ -15,10 +15,18 @@ def run
         # 2. Un-front all pending transaction associated with this donation
         donation.canonical_pending_transactions.update_all(fronted: false)
 
-        # 3. Process remotely
+        # 3. Waive all fees collected
+        donation.canonical_transactions.each do |ct|
+          fee = ct.fee
```
Comment (sampoder) [inline app/services/donation_service/refund.rb]: @garyhtou does that work?

**Review by cjdenio — CHANGES_REQUESTED** [review]
https://github.com/hackclub/hcb/blob/13576eb84d5ff7bf7436bec1600c95869c8df91b/app/services/fee_engine/create.rb also needs to be updated for when a CanonicalTransaction comes in _after_ the donation is refunded

File: app/services/donation_service/refund.rb (hunk 3 above)
Comment (garyhtou) [reply app/services/donation_service/refund.rb]: i believe so

File: app/services/fee_engine/create.rb [hunk 4]
```diff
@@ -47,6 +47,8 @@ def determine_reason
 
       reason = "REVENUE WAIVED" if canonical_transaction.memo.downcase.include?("acctverify") && canonical_transaction.amount_cents.abs < 100 # Waive fees on account verification transactions from platforms like Venmo
 
+      reason = "REFUNDED" if canonical_transaction.donation&.refunded?
```
Comment (sampoder) [inline app/services/fee_engine/create.rb]: @cjdenio would this do the trick?

Comment (cjdenio) [inline app/services/fee_engine/create.rb]: `CanonicalTransaction#donation`, unintuitively, doesn't work for modern donations

```suggestion
      reason = "DONATION REFUNDED" if canonical_transaction.local_hcb_code.donation&.refunded?
```

Comment (sampoder) [reply app/services/fee_engine/create.rb]: Oh neat good catch

**Review by cjdenio — APPROVED** [review]

**Pull request merged**